File: Rezoom.Test/TestCaching.fs
module Rezoom.Test.TestCaching
open Rezoom
open NUnit.Framework
open FsUnit

[<Test>]
let ``strict cached pair`` () =
    {   Task = fun () ->
            plan {
                let! q1 = send "q"
                let! q2 = send "q"
                return q1 + q2
            }
        Batches =
            [   [ "q" ]
            ]
        Result = Good "qq"
    } |> test

[<Test>]
let ``concurrent cached pair`` () =
    {   Task = fun () ->
            plan {
                let! q1, q2 = send "q", send "q"
                return q1 + q2
            }
        Batches =
            [   [ "q" ]
            ]
        Result = Good "qq"
    } |> test

[<Test>]
let ``chaining cached concurrency`` () =
    let testTask x =
        plan {
            let! a = send (x + "1")
            let! b = send (x + "2")
            let! c = send (x + "3")
            return a + b + c
        }
    {   Task = fun () ->
            plan {
                let! x1, x2, x3 =
                    testTask "x", testTask "x", testTask "x"
                return x1 + " " + x2 + " " + x3
            }
        Batches =
            [   [ "x1" ]
                [ "x2" ]
                [ "x3" ]
            ]
        Result = Good "x1x2x3 x1x2x3 x1x2x3"
    } |> test

[<Test>]
let ``still valid after other`` () =
    {   Task = fun () ->
            plan {
                let! q1 = send "q"
                let! q2 = send "q"
                let! m = send "x"
                let! q3 = send "q"
                return q1 + q2 + m + q3
            }
        Batches =
            [   [ "q" ]
                [ "x" ]
            ]
        Result = Good "qqxq"
    } |> test

[<Test>]
let ``invalidation invalidates`` () =
    {   Task = fun () ->
            plan {
                let! q1 = send "q"
                let! q2 = send "q"
                let! m = mutate "x"
                let! q3 = send "q"
                let! q4 = send "q"
                return q1 + q2 + m + q3 + q4
            }
        Batches =
            [   [ "q" ]
                [ "x" ]
                [ "q" ]
            ]
        Result = Good "qqxqq"
    } |> test

[<Test>]
let ``deferred execution`` () =
    let px =
        plan {
            let! q = send "q"
            let! x = send "x"
            return x
        }
    let py = send "y"
    {   Task = fun () ->
            plan {
                let! q = send "q"
                // when px and py are batched together, at first there is a step with both
                // q (from px) and y (from py) pending.
                // q will be pulled from the cache, but rather than just executing y, we should
                // defer y and advance px so we can batch x and y together.
                let! x, y = px, py
                let! z = send "z"
                return q + x + y + z
            }
        Batches =
            [   [ "q" ]
                [ "x"; "y" ] // x and y batched together
                [ "z" ]
            ]
        Result = Good "qxyz"
    } |> test

File: src/Rezoom.CS.Test/TestCaching.cs
using NUnit.Framework;
using Rezoom;
using Rezoom.CS;
using static Rezoom.CS.Test.Helpers;

namespace Rezoom.CS.Test;

[TestFixture]
public class TestCaching
{
    [Test]
    public void Strict_cached_pair()
    {
        async Plan<string> Task()
        {
            var q1 = await Send("q");
            var q2 = await Send("q");
            return q1 + q2;
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "q" } },
            ExpectedResult<string>.Ok("qq"));
    }

    [Test]
    public void Concurrent_cached_pair()
    {
        async Plan<string> Task()
        {
            var (q1, q2) = await Plans.Tuple(Send("q"), Send("q"));
            return q1 + q2;
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "q" } },
            ExpectedResult<string>.Ok("qq"));
    }

    [Test]
    public void Chaining_cached_concurrency()
    {
        async Plan<string> TestTask(string x)
        {
            var a = await Send(x + "1");
            var b = await Send(x + "2");
            var c = await Send(x + "3");
            return a + b + c;
        }
        async Plan<string> Task()
        {
            var (x1, x2, x3) = await Plans.Tuple(TestTask("x"), TestTask("x"), TestTask("x"));
            return x1 + " " + x2 + " " + x3;
        }
        TestRunner.Run(
            Task,
            new[] {
                new[] { "x1" },
                new[] { "x2" },
                new[] { "x3" },
            },
            ExpectedResult<string>.Ok("x1x2x3 x1x2x3 x1x2x3"));
    }

    [Test]
    public void Still_valid_after_other()
    {
        async Plan<string> Task()
        {
            var q1 = await Send("q");
            var q2 = await Send("q");
            var m = await Send("x");
            var q3 = await Send("q");
            return q1 + q2 + m + q3;
        }
        TestRunner.Run(
            Task,
            new[] {
                new[] { "q" },
                new[] { "x" },
            },
            ExpectedResult<string>.Ok("qqxq"));
    }

    [Test]
    public void Invalidation_invalidates()
    {
        async Plan<string> Task()
        {
            var q1 = await Send("q");
            var q2 = await Send("q");
            var m = await Mutate("x");
            var q3 = await Send("q");
            var q4 = await Send("q");
            return q1 + q2 + m + q3 + q4;
        }
        TestRunner.Run(
            Task,
            new[] {
                new[] { "q" },
                new[] { "x" },
                new[] { "q" },
            },
            ExpectedResult<string>.Ok("qqxqq"));
    }

    [Test]
    public void Deferred_execution()
    {
        async Plan<string> Px()
        {
            var q = await Send("q");
            var x = await Send("x");
            return x;
        }
        var py = Send("y");

        async Plan<string> Task()
        {
            var q = await Send("q");
            // When px and py are batched together, at first there is a step with
            // both q (from px) and y (from py) pending. q will be pulled from the
            // cache, but rather than just executing y, we defer y and advance px
            // so we can batch x and y together.
            var (x, y) = await Plans.Tuple(Px(), py);
            var z = await Send("z");
            return q + x + y + z;
        }
        TestRunner.Run(
            Task,
            new[] {
                new[] { "q" },
                new[] { "x", "y" },
                new[] { "z" },
            },
            ExpectedResult<string>.Ok("qxyz"));
    }
}
